src/glewlwyd.c

```diff
@@ -1518,22 +1531,33 @@ void print_help(FILE * output) {
  * handles signal catch to exit properly when ^C is used for example
  * I don't like global variables but it looks fine to people who designed this
  */
-void exit_handler(int signal) {
-  if (signal == SIGQUIT || signal == SIGINT || signal == SIGTERM || signal == SIGHUP) {
-    y_log_message(Y_LOG_LEVEL_INFO, "Glewlwyd - Received close signal");
-  } else if (signal == SIGBUS) {
-    y_log_message(Y_LOG_LEVEL_ERROR, "Glewlwyd - Received bus error signal");
-    exit(1);
-  } else if (signal == SIGSEGV) {
-    y_log_message(Y_LOG_LEVEL_ERROR, "Glewlwyd - Received segmentation fault signal");
-    exit(1);
-  } else if (signal == SIGILL) {
-    y_log_message(Y_LOG_LEVEL_ERROR, "Glewlwyd - Received illegal instruction signal");
-    exit(1);
-  }
-  pthread_mutex_lock(&global_handler_close_lock);
-  pthread_cond_signal(&global_handler_close_cond);
-  pthread_mutex_unlock(&global_handler_close_lock);
+void* signal_thread(void *arg) {
+  sigset_t *sigs = arg;
+  int res, signum;
+
+  while (1) {
+    res = sigwait(sigs, &signum);
+	 if (res) {
+      y_log_message(Y_LOG_LEVEL_ERROR, "Glewlwyd - Waiting for signals failed");
+		exit(1);
+	 }
+    if (signum == SIGQUIT || signum == SIGINT || signum == SIGTERM || signum == SIGHUP) {
+      y_log_message(Y_LOG_LEVEL_INFO, "Glewlwyd - Received close signal: %s", strsignal(signum));
+      pthread_mutex_lock(&global_handler_close_lock);
+      pthread_cond_signal(&global_handler_close_cond);
+      pthread_mutex_unlock(&global_handler_close_lock);
+      return NULL;
+    } else if (signum == SIGBUS) {
+      y_log_message(Y_LOG_LEVEL_ERROR, "Glewlwyd - Received bus error signal");
```

Comment:
Here I'd replace the `y_log_message(Y_LOG_LEVEL_ERROR` with `fprintf(stderr` because `y_log_message`'s `malloc` may cause other problems during a program failure.

Reply:
Yeah, makes sense to keep it simple in these cases.